### DISCUSSION TITLE
collision rules

### DISCUSSION THREAD

**Discussion by dannisliang** [bepu/bepuphysics1#38] [Q&A]
Setting collision rules when two objects collide does not seem to be immediately effective

**Comment by RossNordby** [comment]
The collision rule for a pair is calculated by the broad phase upon pair overlap detection. Modifications to the collision rules of objects after that point won't be seen until the next frame, if I remember correctly.

(Incidentally, this is part of the reason why bepuphysics2 doesn't have 'collision rules', just callbacks that return true/false for creating a pair or creating a constraint that the user defines however they want.)

**Reply by dannisliang** [reply]
How to modify it? implement an immediate rule change ，thank you

**Reply by RossNordby** [reply]
It's been years and years since I messed with this, so I don't remember all the details- turns out there *is* actually an event handler (see the `collisionRulesUpdatedDelegate` on `BroadPhaseEntry`) through which the pair overlaps get updated when collision rules on collidable are modified, and that should affect any already-created pairs, which should in turn inform contact constraints later that frame.

Is it possible that you're using a [deferred event handler](https://github.com/bepu/bepuphysics1/blob/master/Documentation/CollisionEvents.md#2--immediate-versus-deferred-events) that runs after the timestep ends, rather than an immediate event handler?
